tests/FSharp.Compiler.ComponentTests/EmittedIL/RealInternalSignature/ClassTypeInitialization.fs

```diff
@@ -842,6 +847,325 @@ namespace Microsoft.FSharp.Core.CompilerServices
+        ]
+
+    let AssignWithStrength (tok, newData) =
+        let actualLength = List.length newData
+        let tossThreshold = max 0 (actualLength - keepMax) // Delete everything less than this threshold
+        let weakThreshold = max 0 (actualLength - keepStrongly) // Weaken everything less than this threshold
+
+        let newData = newData |> List.mapi (fun n kv -> n, kv) // Place the index.
+
+        let newData =
+            newData
+            |> List.filter (fun (n: int, v) -> n >= tossThreshold || requiredToKeep (snd v))
+
+        let newData =
+            newData
+            |> List.map (fun (n: int, (k, v)) ->
+                let handle =
+                    if n < weakThreshold && not (requiredToKeep v) then
+                        Weak(WeakReference<_>(v))
+                    else
+                        Strong(v)
+
+                k, handle)
+
+        ignore tok // Updating refs requires tok
+        refs <- newData
+
+    member al.TryPeekKeyValue(tok, key) =
+        // Returns the original key value as well since it may be different depending on equality test.
+        let data = FilterAndHold(tok)
+        TryPeekKeyValueImpl(data, key)
+
+    member al.TryGetKeyValue(tok, key) =
+        let data = FilterAndHold(tok)
+        let result, newData = TryGetKeyValueImpl(data, key)
+        AssignWithStrength(tok, newData)
+        result
+
+    member al.TryGet(tok, key) =
+        let data = FilterAndHold(tok)
+        let result, newData = TryGetKeyValueImpl(data, key)
+        AssignWithStrength(tok, newData)
+
+        match result with
+        | Some(_, value) -> Some(value)
+        | None -> None
+
+    member al.Put(tok, key, value) =
+        let data = FilterAndHold(tok)
+
+        let data = if Exists(data, key) then RemoveImpl(data, key) else data
+
+        let data = Add(data, key, value)
+        AssignWithStrength(tok, data) // This will remove extras
+
+    member al.Remove(tok, key) =
+        let data = FilterAndHold(tok)
+        let newData = RemoveImpl(data, key)
+        AssignWithStrength(tok, newData)
+
+    member al.Clear(tok) =
+        let _discards = FilterAndHold(tok)
+        AssignWithStrength(tok, [])
+
+    member al.Resize(tok, newKeepStrongly, ?newKeepMax) =
+        let newKeepMax = defaultArg newKeepMax 75
+        keepStrongly <- newKeepStrongly
+        keepMax <- max newKeepStrongly newKeepMax
+        let keep = FilterAndHold(tok)
+        AssignWithStrength(tok, keep)
+    """
+        |> asLibrary
+        |> withRealInternalSignature realSig
+        |> withOptimization
+        |> compile
+        |> shouldSucceed
+
+    [<InlineData(true, true)>]          // RealSig Optimize
+    [<InlineData(true, false)>]         // RealSig NoOptimize
+    [<InlineData(false, true)>]         // Regular Optimize
+    [<InlineData(false, false)>]        // Regular NoOptimize
+    [<Theory>]
+    let ``BigTuples`` (realSig, optimize) =
+        let withOptimization compilation =
```

Comment:
This block is repeated across tests, can be moved on top of the file:

`|> withRealSig realSig |> withOptimize optimize`

Reply:
I can add something to CompilerUtilities, it is general enough

Reply:
Done